### PR TITLE
fix ipBlock except for network policy

### PR DESCRIPTION
**Pull request by JacobTanenbaum**
our current implementation of Ipblock is incorrect. creating a deny rule for the
except blocks will not allow ips that overlap another IPblock to be allowed.

I changed this so that we only create one rule per ipBlock

so if you have one policy that says

```
- from:
  - ipBlock:
      cidr: 10.0.0.0/8
      except:
      - 10.0.1.0/24

```
and another that says

```
- from:
  - ipBlock:
      cidr: 10.0.0.0/8

```
it will translate two acls that are

```
match=ip4.src == 10.0.0.0/8 && ip4.src !={10.0.1.0/24}
match=ip4.src == 10.0.0.0/8

```
the first rule will not catch a packet from 10.0.1.0 it will
cascade to the second ACL and correctly be allowed. If the except
is not overlapping an additional range it will be denied by the
default deny ACL

This should resolve https://github.com/ovn-org/ovn-kubernetes/issues/845

Signed-off-by: Jacob Tanenbaum <jtanenba@redhat.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

I included some unit testing and enbabled the features e2e test. 

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/ovn/gress_policy.go [hunk 1]
```diff
@@ -65,6 +71,10 @@ func (pp *portPolicy) getL4Match() (string, error) {
 }
 
 func newGressPolicy(policyType knet.PolicyType, idx int, namespace, name string) *gressPolicy {
+	ipBlock := ipBlock{
+		except: make([]string, 0),
```
Comment (dcbw) [inline go-controller/pkg/ovn/gress_policy.go:75]: @JacobTanenbaum I think you can just leave this empty; a nil array should work fine here. In fact, you can just do `&ipBlock{},` below.

File: go-controller/pkg/ovn/gress_policy.go [hunk 2]
```diff
@@ -382,79 +388,38 @@ func (gp *gressPolicy) modifyACLAllow(oldMatch, newMatch string) error {
...
 			v4CIDRs = append(v4CIDRs, cidr)
 		}
 	}
+
 	if len(v4CIDRs) > 0 {
 		v4AddressSetStr := strings.Join(v4CIDRs, ", ")
-		v4MatchStr = fmt.Sprintf("%s.%s == {%s}", "ip4", direction, v4AddressSetStr)
-		matchStr = v4MatchStr
+		v4MatchStr = fmt.Sprintf("%s.%s != {%s}", "ip4", direction, v4AddressSetStr)
+		matchStr = fmt.Sprintf("%s && %s", matchStr, v4MatchStr)
```
Comment (aojea) [inline go-controller/pkg/ovn/gress_policy.go:413]: should we need parenthesis to concatenate expressions?

File: go-controller/pkg/ovn/gress_policy.go [hunk 3]
```diff
@@ -128,6 +139,9 @@ func (gp *gressPolicy) addPortPolicy(portJSON *knet.NetworkPolicyPort) {
 }
 
 func (gp *gressPolicy) addIPBlock(ipblockJSON *knet.IPBlock) {
+	gp.ipBlock.cidr = ipblockJSON.CIDR
+	gp.ipBlock.except = append(gp.ipBlock.except, ipblockJSON.Except...)
```
Comment (danwinship) [inline go-controller/pkg/ovn/gress_policy.go:143]: `addIPBlock()` can be called multiple times on a given `gressPolicy`, which is why you need an array of `ipBlock`. (A `gressPolicy` represents a single `knet.NetworkPolicyIngressRule`/`knet.NetworkPolicyEgressRule`, which can have multiple `NetworkPolicyPeer`s, each with an `ipBlock`.)

And then below you will need to generate a separate ACL for each `ipBlock` in the policy.

File: go-controller/pkg/ovn/gress_policy.go [hunk 4]
```diff
@@ -37,6 +37,12 @@ type gressPolicy struct {
 	// except the IP block in the except, which should be dropped.
 	ipBlockCidr   []string
 	ipBlockExcept []string
+	ipBlock       *ipBlock
+}
+
+type ipBlock struct {
+	cidr   string
+	except []string
```
Comment (danwinship) [inline go-controller/pkg/ovn/gress_policy.go:45]: Really you don't even need this type; you can just use `knet.IPBlock` directly instead

File: go-controller/pkg/ovn/gress_policy.go (hunk 1 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

File: go-controller/pkg/ovn/gress_policy.go [hunk 5]
```diff
@@ -37,6 +37,12 @@ type gressPolicy struct {
 	// except the IP block in the except, which should be dropped.
 	ipBlockCidr   []string
 	ipBlockExcept []string
+	ipBlock       *ipBlock
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

File: go-controller/pkg/ovn/gress_policy.go (hunk 3 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

File: go-controller/pkg/ovn/gress_policy.go (hunk 4 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

File: go-controller/pkg/ovn/gress_policy.go [hunk 6]
```diff
@@ -382,79 +388,38 @@ func (gp *gressPolicy) modifyACLAllow(oldMatch, newMatch string) error {
...
-	for _, cidr := range ipCIDRs {
+	if utilnet.IsIPv6CIDRString(ipBlock.cidr) {
+		matchStr = fmt.Sprintf("%s.%s == %s", "ip6", direction, ipBlock.cidr)
+	} else {
+		matchStr = fmt.Sprintf("%s.%s == %s", "ip4", direction, ipBlock.cidr)
+	}
+
+	for _, cidr := range ipBlock.except {
 		if utilnet.IsIPv6CIDRString(cidr) {
 			v6CIDRs = append(v6CIDRs, cidr)
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

File: go-controller/pkg/ovn/gress_policy.go (hunk 2 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: it can make the logic easier to read in my opinion

File: go-controller/pkg/ovn/gress_policy.go [hunk 7]
```diff
@@ -382,79 +373,45 @@ func (gp *gressPolicy) modifyACLAllow(oldMatch, newMatch string) error {
...
+		v6ExceptCIDRs := make([]string, 0)
+		var exceptStr string
+		if utilnet.IsIPv6CIDRString(ipBlock.CIDR) {
+			matchStr = fmt.Sprintf("%s.%s == %s", "ip6", direction, ipBlock.CIDR)
 		} else {
-			v4CIDRs = append(v4CIDRs, cidr)
+			matchStr = fmt.Sprintf("%s.%s == %s", "ip4", direction, ipBlock.CIDR)
+		}
+		for _, exceptIP := range ipBlock.Except {
+			if utilnet.IsIPv6CIDRString(exceptIP) {
```
Comment (danwinship) [inline go-controller/pkg/ovn/gress_policy.go:389]: each except IP has to be a subset of the top-level CIDR, which implies that all except IPs for a given cidr will be the same IP family as that cidr. (This is enforced by apiserver validation.) So the overall `matchStr` will be one of `ip4.DIR == CIDR`, `ip6.DIR == CIDR`, `ip4.DIR == CIDR && ip4.DIR != { EXCEPTS }` or `ip6.DIR == CIDR && ip6.DIR != { EXCEPTS }`

File: go-controller/pkg/ovn/gress_policy.go [hunk 8]
```diff
@@ -65,6 +64,7 @@ func (pp *portPolicy) getL4Match() (string, error) {
 }
 
 func newGressPolicy(policyType knet.PolicyType, idx int, namespace, name string) *gressPolicy {
+
```
Comment (danielmellado) [inline go-controller/pkg/ovn/gress_policy.go:67]: As Dan comments, please remove this one (if only to be consistent with the same style along the file)

**Review by danielmellado — COMMENTED** [review]
Besides the last comment from Dan, the overall code looks cleaner now. Thanks Jacob!

File: go-controller/pkg/ovn/gress_policy.go [hunk 9]
```diff
@@ -35,8 +35,7 @@ type gressPolicy struct {
 
 	// ipBlockCidr represents the CIDR from which traffic is allowed
 	// except the IP block in the except, which should be dropped.
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

File: go-controller/pkg/ovn/gress_policy.go (hunk 7 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: did not realize that it was enforced that they had to be a subset of the CIDR block, fixed

File: go-controller/pkg/ovn/gress_policy.go (hunk 8 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

File: go-controller/pkg/ovn/gress_policy.go [hunk 10]
```diff
@@ -73,8 +70,6 @@ func newGressPolicy(policyType knet.PolicyType, idx int, namespace, name string)
 		peerV4AddressSets: sets.String{},
 		peerV6AddressSets: sets.String{},
 		portPolicies:      make([]*portPolicy, 0),
-		ipBlockCidr:       make([]string, 0),
-		ipBlockExcept:     make([]string, 0),
```
Comment (abhat) [inline go-controller/pkg/ovn/gress_policy.go:77]: nit: why not make the slice here to be sure?

File: go-controller/pkg/ovn/gress_policy.go [hunk 11]
```diff
@@ -186,28 +180,24 @@ func (gp *gressPolicy) sizeOfAddressSet() int {
...
-		} else {
-			match = fmt.Sprintf("match=\"%s && %s && %s\"",
-				ipBlockMatch, l4Match, lportMatch)
-		}
+		ipBlockMatches = constructIPBlockStringsForACL("src", gp.ipBlock)
 	} else {
-		ipBlockMatch := constructIPStringForACL("dst", gp.ipBlockCidr)
+		ipBlockMatches = constructIPBlockStringsForACL("dst", gp.ipBlock)
+	}
+	for i, ipBlockMatch := range ipBlockMatches {
```
Comment (abhat) [inline go-controller/pkg/ovn/gress_policy.go:190]: I am curious as to why we are not doing this for loop inside the constructIPBlockStringsForACL itself.

File: go-controller/pkg/ovn/gress_policy.go (hunk 10 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go:77]: you are guaranteed to call addIPBlock before generating the match string

File: go-controller/pkg/ovn/gress_policy.go (hunk 11 above)
Comment (JacobTanenbaum) [reply go-controller/pkg/ovn/gress_policy.go]: done

**Comment by abhat** [comment]
LGTM.

**Pull request merged**